### ISSUE TITLE
Find contact by typing in username or name

### ISSUE THREAD

**Issue by simison** [Trustroots/trustroots#530]
Show a list of members by typing in *username*, *firstName*, *lastName* or *displayName*.

- The API is already implemented at #531
- Markup can be basically the same as at contacts tab on the user profile
- First iteration doesn't need pagination, just hard limit (e.g. 10-30 or something)
- ~First iteration doesn't need to find from the middle of the name, only from the beginning~ We'll not implement partial search at all, see https://github.com/Trustroots/trustroots/pull/531#issuecomment-430018690
- Should not be case sensitive
- In future this page would hold also FB your friends who are also on TR (#392).
- Might need adding another Mongo index to *firstName*, *lastName* and *displayName*.

In navigation - on mobile:
![image](https://user-images.githubusercontent.com/87168/28264027-c2694742-6af1-11e7-8f1c-90c5cc892300.png)

In navigation - on desktop:
![image](https://user-images.githubusercontent.com/87168/28264058-e2440e58-6af1-11e7-842d-9812761e843e.png)

Label could be "Find friends".

The goal of this feature is:
- to find your existing contacts easier from Trustroots so that you can 1) add them as contacts 2) write to them.
- to not to find hosts or members to meet.

**Comment by danidanitr** [comment]
We continue with this tomorrow :)

**Comment by danidanitr** [comment]
So far:

**left side: icon of contacts + plus
right side: search through the contact list (shorter: search your contacts)
contact count (+pending) = centered
results = users appear with the same design**

![contacts_1](https://user-images.githubusercontent.com/43136016/47148191-5d96e200-d2d0-11e8-946d-43419d4edc36.jpg)

Then, when you type in "add new contacts..."
![contacts_2](https://user-images.githubusercontent.com/43136016/47148678-984d4a00-d2d1-11e8-8baf-1482fec77be4.jpg)

**Comment by danidanitr** [comment]
So far:

left side: icon of contacts + plus
right side: search through the contact list (shorter: search your contacts)
contact count (+pending) = centered
results = users appear with the same design
![contacts_1](https://user-images.githubusercontent.com/43136016/47148972-8029fa80-d2d2-11e8-9464-b91ee7889a36.jpg)

Then, when you type in "add new contacts..."
![contacts_2](https://user-images.githubusercontent.com/43136016/47148983-861fdb80-d2d2-11e8-8856-f97d2f11ea2c.jpg)

**Comment by danidanitr** [comment]
So far:

left side: icon of contacts + plus
right side: search through the contact list (shorter: search your contacts)
contact count (+pending) = centered
results = users appear with the same design
![contacts_1](https://user-images.githubusercontent.com/43136016/47149686-946ef700-d2d4-11e8-85a6-7f9f168ba4eb.jpg)

Then, when you type in "add new contacts..."
![contacts_2](https://user-images.githubusercontent.com/43136016/47149691-989b1480-d2d4-11e8-88c0-9234142bf097.jpg)

**Comment by simison** [comment]
Would we have any uses cases for this search outside of "adding them as a contact" -context?

The search now suggests that anyone showing up will be added as a contact — how? An extra button?

**Comment by Akronix** [comment]
For me, going to your contacts to find new people (or people you already know from outside trustroots) in trustroots is counter-intuitive.
The search should be another thing outside contacts and closer related to the "Search" tab. I see that other social networks (browser, etc.) have one only search bar, or search place. Users go there, type words of a widely spectrum of different approaches to find (username, name, description, interests, etc.) people or an event, or an organization or a place, etc and get results grouped by their different type (people, events, places, etc.)

I agree with starting with something simple so maybe just people and username and names, but I'd definitely keep that separated from contacts and I'd try to match it better with the current Search feature.

**Comment by simison** [comment]
In very long vision we'll probably end up having "one master search" field at the header of the page all the time visible, but we'll need to think in steps how to get there.

If to expose this feature via existing contacts tab, to add a new page, or implementing both is a good question to think about, though.

**Comment by Akronix** [comment]
hmm I think I got confused, so actually #721 it is what I was asking about. Then #705 is then a duplicate of #721 and not to this.

Well, as you said @simison a probably second round of thought will be necessary at some point in order to integrate better everything.

**Comment by simison** [comment]
Gotcha 👍  Plenty of space for confusing there indeed!

I'd say improving both searches this way is equally high on priority at this point. "Finding people you already know" is probably easier since map-search code can be pretty complicated.

**Comment by simison** [comment]
Can you list some possible scenarios here? Would be useful.

**Comment by danidanitr** [comment]
Although I agree with being able to search from a main search panel, this is for now the quicker solution to solve this impasse. To avoid mixing "searching for adding" and "searching for browsing" then we can just not make obvious the fact you are searching to add someone, simply searching, leaving the decision to the user whether they open the profile by clicking on it or they only click some kind of green button (easy to spot) which would simply add that person. I think it is good when the user can choose and we create less.

**Comment by Regenwasser** [comment]
> Can you list some possible scenarios here? Would be useful.

Often they might just want to **check out their profile** or want to **contact them**. So, on top of my head the first ones:

1. Two friends surf together and send a message like: "I'm travelling with _Hitchmaster_ - check out his profile". Then the user throws that name into the member search to look at both profiles before making a decision.

2. A user briefly met someone at a Trustroots meeting or somewhere else, but forgot to exchange details and now would like to send them a message.

3. The user is at an event and met a Trustroots member that offers to host him. He wants to check out his profile before.

4. Before a trip a users meet a friend that tells him "Last year I used Trustroots to stay with a guy called Michael in Kaunas", so he enters name into the member search to see if he still lives there.

It seems we'd miss out on all those use cases, because people won't think of adding someone and are also building this only for an edge case for those two reasons:

- Trustroots is a small community and it's _(for the average user - who doesn't have hundreds of CS contacts or is active on our Slack)_ not as likely that many of his friends are on Trustroots. So, wanting to add many contacts/friends isn't something that often happens.  

- Trustroots not a new cool messenger or a social network where you sign up and want to add all your contacts. So far there are no benefits yet to have many contacts here. If you sign up to Facebook you add your contacts to see their stuff in your timeline and stay in touch, on Instagram you follow someone to see their pictures and on Snapchat you add someone to see their Snaps.

On here you mainly add someone as a contact after you met them through hosting/surfing and in that case you already know their profile and add them through that.

**Comment by danidanitr** [comment]
Search form changed to a more general approach
![adding_2](https://user-images.githubusercontent.com/43136016/47167919-de210700-d2ff-11e8-9f9f-5189a31693c1.jpg)

When finding user I choose if to click on the profile card and them jump to the profile or simply clicking that corner + square, I add the user and jump to that page where it says "do you wanna add as your friend" and then that's it. 2x1
![adding_1](https://user-images.githubusercontent.com/43136016/47167934-e8430580-d2ff-11e8-8269-680207a9ebcc.jpg)

**Comment by danidanitr** [comment]
How do you see this fast mockup working according to what you say? I personally like the idea users can decide what to do.

**Comment by Regenwasser** [comment]
What you did very looks good. What I unsuccessfully tried to illustrate a few times on Slack (as well as here today) is that the member search needs to be directly accessible for everyone and that searching contacts isn't useful enough.

So, because we decided that we can't implement it correctly right now and still have features like #721 (and high priority stuff) to go before we can do that I believe we should put those sweet developer skills to good use elsewhere. 

Explaining why this is "[Priority] Low", or rather "Won't do" for me, is my contribution :). What I'd see working best is to close "Find _contact_ by typing in username or name" and open "Find _member_ by typing in username or name".

**Comment by guaka** [comment]
I'd like to add my use case, which is not (nor going to be) very common, but I think it's a very beneficial use case to get more people involved long term.

I want the query to match at least these places
- `username`
- `displayName`, `firstName`, `lastName`
- `email`
- github username

Optional:
- other networks apart from github
- go through the entire user doc recursively, e.g. from Facebook we have `first_name`, `last_name` and potentially another email address

I want it when I want to know the person behind a commit, behind a support request, or behind a message on slack.

This doesn't need any fancy graphics, it doesn't need to be in the menu. If I have this functionality today it will make my life easier because I won't need to run custom mongo queries.

(Generally I think it can be good if we allow for some experimentation, where some new, working yet unfinished features are accessible by going directly through a specific URL.  Later also through a "beta-tester" role to some folks.)
